File: frontend/flutter_application_desktop/lib/models/product.dart
import 'enum/estado_producto.dart';

class Product {
  final int? idProduct;
  final String? codigo;
  final String nombre;
  final String contenido;
  final double precio;
  final String marca;
  final String? fotoUrl;
  final EstadoProducto estado;
  final DateTime? fechaCreacion;
  final DateTime? fechaActualizacion;
  final int totalVendido;
  final int? categoriaId;
  final int? clasificacionId;

  Product({
    this.idProduct,
    this.codigo,
    required this.nombre,
    required this.contenido,
    required this.precio,
    required this.marca,
    this.fotoUrl,
    this.estado = EstadoProducto.ACTIVO,
    this.fechaCreacion,
    this.fechaActualizacion,
    this.totalVendido = 0,
    required this.categoriaId,
    required this.clasificacionId,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      idProduct: json['idProduct'],
      codigo: json['codigo'],
      nombre: json['nombre'],
      contenido: json['contenido'],
      precio: (json['precio'] as num).toDouble(),
      marca: json['marca'],
      fotoUrl: json['fotoUrl'],
      estado: estadoFromString(json['estado']),
      fechaCreacion: json['fechaCreacion'] != null
          ? DateTime.parse(json['fechaCreacion'])
          : null,
      fechaActualizacion: json['fechaActualizacion'] != null
          ? DateTime.parse(json['fechaActualizacion'])
          : null,
      totalVendido: json['totalVendido'] ?? 0,
      categoriaId: json['categoria'] != null
          ? json['categoria']['idCategoria']
          : null,
      clasificacionId: json['clasificacion'] != null
          ? json['clasificacion']['idClasificacion']
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      "codigo": codigo,
      "nombre": nombre,
      "contenido": contenido,
      "precio": precio,
      "marca": marca,
      "fotoUrl": fotoUrl,
      "fechaCreacion": fechaCreacion,
      "fechaActualizacion": fechaActualizacion,
      "estado": estadoToString(estado),
      "categoriaId": categoriaId,
      "clasificacionId": clasificacionId,
    };
  }

  Product copyWith({EstadoProducto? estado, int? totalVendido}) {
    return Product(
      idProduct: idProduct,
      codigo: codigo,
      nombre: nombre,
      contenido: contenido,
      precio: precio,
      marca: marca,
      fotoUrl: fotoUrl,
      estado: estado ?? this.estado,
      fechaCreacion: fechaCreacion,
      fechaActualizacion: fechaActualizacion,
      totalVendido: totalVendido ?? this.totalVendido,
      categoriaId: categoriaId,
      clasificacionId: clasificacionId,
    );
  }
}

File: backend/springboot_backend_desktop/src/main/java/com/project/flutter_backend_desktop/modelo/Product.java
package com.project.flutter_backend_desktop.modelo;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.project.flutter_backend_desktop.modelo.Enum.EstadoProducto;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Data
@NoArgsConstructor
@AllArgsConstructor
@ToString()
@JsonInclude(JsonInclude.Include.NON_DEFAULT)
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    Integer idProduct;

    @Column(nullable = true, unique = true)
    String codigo;

    @NotBlank(message = "El nombre es obligatorio")
    String nombre;
    String contenido;

    @Positive(message = "El precio debe ser mayor a 0")
    Double precio;
    String marca;
    String fotoUrl;

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    private EstadoProducto estado;

    @CreationTimestamp
    private LocalDateTime fechaCreacion;

    @UpdateTimestamp
    private LocalDateTime fechaActualizacion;

    private Integer totalVendido = 0;

    @NotNull(message = "Debe seleccionar una categoría")
    @ManyToOne
    @JoinColumn(name = "categoria_id")
    private Categoria categoria;

    @NotNull(message = "Debe seleccionar una clasificacion")
    @ManyToOne
    @JoinColumn(name = "clasificacion_id")
    private Clasificacion clasificacion;

    @OneToMany(mappedBy = "productoOrigen", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<PromoBonoProduct> bonificaciones = new ArrayList<>();
}